src/system/random/basic.lean

```diff
@@ -0,0 +1,467 @@
+def random_r [preorder α] [bounded_random α] (x y : α) (h : x ≤ y) : rand_g g (x .. y) :=
+bounded_random.random_r g x y h
+
+/-- generate an infinite series of random values of type `α` between `x` and `y` -/
+def random_series_r [preorder α] [bounded_random α] (x y : α) (h : x ≤ y) : rand_g g (stream (x .. y)) :=
+do gen ← uliftable.up (split g),
+   pure $ corec_state (bounded_random.random_r g x y h) gen
+
+end rand
+
+namespace io
+
+private def accum_char (w : ℕ) (c : char) : ℕ :=
+c.to_nat + 256 * w
+
+/-- create and a seed a random number generator -/
+def mk_generator : io std_gen := do
+seed ← io.rand 0 shift_31_left,
+return $ mk_std_gen seed
+
+variables {α : Type}
+
+/-- run `cmd` using a randomly seeded random number generator -/
+def run_rand (cmd : _root_.rand α) : io α :=
+do g ← io.mk_generator,
+   return $ (cmd.run ⟨g⟩).1
+
+section random
+variables [preorder α] [random α]
+
+/-- randomly generate a value of type α -/
+def random : io α :=
+io.run_rand (rand.random α)
+
+/-- randomly generate an infinite series of value of type α -/
+def random_series : io (stream α) :=
+io.run_rand (rand.random_series α)
+
+end random
+
+section bounded_random
+variables [preorder α] [bounded_random α]
+
+/-- randomly generate a value of type α between `x` and `y` -/
+def random_r (x y : α) (p : x ≤ y) : io (x .. y) :=
+io.run_rand (bounded_random.random_r _ x y p)
+
+/-- randomly generate an infinite series of value of type α between `x` and `y` -/
+def random_series_r (x y : α) (h : x ≤ y) : io (stream $ x .. y) :=
+io.run_rand (rand.random_series_r x y h)
+
+end bounded_random
+
+end io
+
+namespace tactic
+
+/-- create a seeded random number generator in the `tactic` monad -/
+meta def mk_generator : tactic std_gen := do
+tactic.unsafe_run_io @io.mk_generator
+
+/-- run `cmd` using the a randomly seeded random number generator
+in the tactic monad -/
+meta def run_rand {α : Type u} (cmd : rand α) : tactic α := do
+⟨g⟩ ← tactic.up mk_generator,
+return (cmd.run ⟨g⟩).1
+
+variables {α : Type u}
+
+section bounded_random
+variables [preorder α] [bounded_random α]
+
+/-- use `random_r` in the `tactic` monad -/
+meta def random_r (x y : α) (h : x ≤ y) : tactic (x .. y) :=
+run_rand (rand.random_r x y h)
+
+/-- use `random_series_r` in the `tactic` monad -/
+meta def random_series_r (x y : α) (h : x ≤ y) : tactic (stream $ x .. y) :=
+run_rand (rand.random_series_r x y h)
+
+end bounded_random
+
+section random
+
+variables [preorder α] [random α]
+
+/-- use `random` in the `tactic` monad -/
+meta def random : tactic α :=
+run_rand (rand.random α)
+
+/-- use `random_series` in the `tactic` monad -/
+meta def random_series : tactic (stream α) :=
+run_rand (rand.random_series α)
+
+end random
+
+end tactic
+
+namespace bool
+
+/-- Make `i` into an element of the interval `x .. y` if feasible
+and return an arbitrary element of `x .. y` otherwise -/
+def fit_in_Icc (x y : bool) (p : x ≤ y) (i : bool) : x .. y := do
```

Comment:
The name and documentation are relatively generic. I would like to see something indicating why this is specific to `bool` (if `i` is not in `x..y`, then `x = y` so the arbitrary element must be `x`). Perhaps the following, although I am not completely convinced:
```suggestion
/-- Make `i` into an element of the interval `x .. y` if feasible.
Otherwise return `x`, which is the only element of `x .. y` in that case.  -/
def fit_in_Icc (x y : bool) (p : x ≤ y) (i : bool) : x .. y := do
```